auth0/src/main/java/com/auth0/android/provider/WebAuthProvider.java

```diff
@@ -396,30 +446,50 @@ private void startAuthorization(Activity activity, Uri authorizeUri, int request
             Log.d(TAG, "About to start the authorization using the WebView");
             intent = new Intent(activity, WebAuthActivity.class);
             intent.setData(authorizeUri);
-            intent.putExtra(WebAuthActivity.CONNECTION_NAME_EXTRA, connectionName);
+            intent.putExtra(WebAuthActivity.CONNECTION_NAME_EXTRA, getConnection());
             intent.putExtra(WebAuthActivity.FULLSCREEN_EXTRA, useFullscreen);
             activity.startActivityForResult(intent, requestCode);
         }
     }
 
     @VisibleForTesting
-    boolean shouldUsePKCE() {
-        return useCodeGrant && PKCE.isAvailable();
+    static Credentials mergeCredentials(Credentials urlCredentials, Credentials codeCredentials) {
+        final String idToken = codeCredentials.getIdToken() != null ? codeCredentials.getIdToken() : urlCredentials.getIdToken();
+        final String accessToken = codeCredentials.getAccessToken() != null ? codeCredentials.getAccessToken() : urlCredentials.getAccessToken();
+        final String type = codeCredentials.getType() != null ? codeCredentials.getType() : urlCredentials.getType();
+        final String refreshToken = codeCredentials.getRefreshToken() != null ? codeCredentials.getRefreshToken() : urlCredentials.getRefreshToken();
+
+        return new Credentials(idToken, accessToken, type, refreshToken);
     }
 
-    private Uri buildAuthorizeUri() {
+    @VisibleForTesting
+    static boolean hasValidNonce(String nonce, String token) {
+        try {
+            final JWT idToken = new JWT(token);
+            final Claim nonceClaim = idToken.getClaim(KEY_NONCE);
+            return !(nonceClaim == null || !nonce.equals(nonceClaim.asString()));
+        } catch (DecodeException e) {
+            return false;
+        }
+    }
+
+    private boolean shouldUsePKCE() {
+        return getResponseType().contains(RESPONSE_TYPE_CODE) && PKCE.isAvailable();
+    }
+
+    @VisibleForTesting
+    Uri buildAuthorizeUri() {
         final Uri authorizeUri = Uri.parse(account.getAuthorizeUrl());
         String redirectUri = helper.getCallbackURI(account.getDomainUrl());
+        final Map<String, String> queryParameters = new HashMap<>(parameters);
 
-        final Map<String, String> queryParameters = new HashMap<>();
-        queryParameters.put(KEY_CONNECTION_SCOPE, connectionScope);
-        queryParameters.put(KEY_RESPONSE_TYPE, RESPONSE_TYPE_TOKEN);
-
+        if (!getResponseType().contains(RESPONSE_TYPE_ID_TOKEN)) {
+            queryParameters.remove(KEY_NONCE);
```

Comment:
We should only do this if we generated the nonce, if the user provided one we should still send it